Implement feature #7: Calculate individual expenses

BirkLauritzen opened the pull request:
This is a pull request with a contribution that fixes issue 7 and tests for it as well.

This PR solves https://github.com/ST-ITM-2025-2/expense-splitter/issues/7

Implemented the CSV file writing logic in the ExpenseStore append method.
Implemented the expenseToCSV method to convert an Expense object into a CSV string.
Added the ReportTest file
How to test: Run the program -> input 6 and enter -> the amount will now be divided between the participants

happy_path:
Correctly dividing the amount with the participants

edge_case:
empty participant list
empty expense amount
Closes https://github.com/ST-ITM-2025-2/expense-splitter/issues/7

hayat711 commented:
**Group8** will be reviewing this PR [code implemented by group 9 for feature #7 ]

saschahuberzh commented on src/main/java/itm/oss/splitter/Reports.java:
```diff
@@ -1,16 +1,74 @@
+                if (map.get(name) == null) {
+                    map.put(name, new PersonSummary(BigDecimal.ZERO, BigDecimal.ZERO, BigDecimal.ZERO));
+                }
+                PersonSummary ps = map.get(name);
```
Consider renaming ps to something more descriptive like personSummary, it would make the loop logic easier to follow

Poppaluu replied:
ps renamed to personSummary

saschahuberzh commented on src/main/java/itm/oss/splitter/ExpenseStore.java:
```diff
@@ -1,15 +1,52 @@
+        // Format: date,payer,amount,currency,participants,category,notes
+        // participants are semicolon-separated.
+        String line = "";
+        ArrayList<Expense> expenses = new ArrayList<>();
```
Using the List interface instead of ArrayList in the method signature would make the code more flexible and easier to maintain in the future

Poppaluu replied:
ArrayLists now use the List interface

MHJeong730 commented on src/main/java/itm/oss/splitter/Reports.java:
```diff
@@ -1,16 +1,74 @@
+            }
+        }
+
+        for (String name : map.keys()) {
```
I can't figure out what this for loop is doing. It seems to simply add 0 to owedTotal. Is there a specific function intended for this for loop? Could you please add some comments on what this loop does?

Driedoutjerky replied:
The for-loop code you've mentioned adds `BigDecimal.ZERO` to the `owedTotal` for each name in the `personSummary`.
That ensures `recalculateNet()` method to run for the final time for each participant before returning `map` to ensure that no stale numbers have been left un-calculated. @Poppaluu will be adding the comment to the code soon. Thank you for your feedback!

Poppaluu replied:
comments added

MHJeong730 commented on src/main/java/itm/oss/splitter/Reports.java:
```diff
@@ -1,16 +1,74 @@
+                    ps.addPaid(amount);
+                }
+
+                // credit to Kasper (github: kasperB2004) for helping to fix the cent rounding problem
```
I think this idea of ​​solving the problem of sharing indivisible expenses is excellent.

Poppaluu replied:
ty

MHJeong730 commented on src/test/java/itm/oss/splitter/ReportsTest.java:
```diff
@@ -0,0 +1,114 @@
+        assertEquals(3, xs.get(0).getParticipants().size(), "There should be three participants");
+    }
+
+// Unit testing
```
Adding test cases for various edge cases is very helpful. However, it would be better if there were test cases for the getNet() function to ensure that the recalculateNet() function works properly.

BirkLauritzen replied:
I've added two more edge cases, and test cases for getNet as well

jobregon1212 commented:
Hello guys, this PR should only modify the .java files (implementation and test files). It should not touch .gitignore , README, or any other file. So please fix that. Remember to synchronize with upstream repository and pull the latest changes from the upstream repository.

BirkLauritzen committed:
Implement feature #7 (Java-only). Remove non-Java changes from PR.

BirkLauritzen committed:
Update ReportsTest.java

Improve ReportsTest coverage and edge-case validation

Key changes:
Added integration tests for loading expenses.sample.csv
Added multiple in-memory edge-case tests

BirkLauritzen commented:
All the reviewed changes have been implemented and resolved

sengA7 commented on src/test/java/itm/oss/splitter/ReportsTest.java:
```diff
@@ -0,0 +1,199 @@
+    @Test
+    void perPerson_usesCsvLoadedByExpenseStore() throws IOException, URISyntaxException {
+
+        // Load the CSV. Prefer classpath (src/test/resources), fallback to repo file (./data/expenses.sample.csv)
```
Hello, it's Agnes sorry i do the review part too late
but here the code to load expenses.sample.csv is duplicated two times with (perPerson_usesCsvLoadedByExpenseStore and perPerson_netFromSimpleCsv)
The current test coverage is excellent

BirkLauritzen replied:
Thank you for spotting it. The issue has been resolved now

sengA7 commented on src/test/java/itm/oss/splitter/ReportsTest.java:
```diff
@@ -0,0 +1,199 @@
+                .add(m.get("Cara").getNet());
+            assertEquals(0, sum.compareTo(BigDecimal.ZERO), "Sum of net must be zero");
+        }
+}
```
I have nothing to tell more, your code is very clean and organized (easy to read and undersand). Good job guys !!

BirkLauritzen replied:
Thanks!

BirkLauritzen committed:
Update ReportsTest.java

Made changes to the file:
Removed comments.
Added display name to the "perPerson_usesCsvLoadedByExpenseStore" test.
Removed duplicate path finder for the csv file.
Made a path finder helper to the csv file.
Removed test that caused errors.

BirkLauritzen committed:
Update ReportsTest.java

Fixed spelling error

BirkLauritzen committed:
Fixed merge conflicts

I fixed merge conflicts in ExpenseStore: where we're using the current solution from the repository.

Reports.java: where we've successfully merged our solution issue with the group we're sharing our file with.

ReportsTest: where we've also successfully merged our tests together with the group we share the Reports file with.

Also changed the data file to retrieve data from the data/expenses.sample.csv, where we used a dummy file before

BirkLauritzen committed:
Merge branch 'main' into main

BirkLauritzen commented:
All files have been successfully merged with the files from main
